src/ring_theory/graded_ring/basic.lean

```diff
@@ -0,0 +1,244 @@
+  exact hi,
+end⟩
+
+lemma graded_ring.as_sum (r : R) :
+  r = ∑ i in graded_ring.support R A r, graded_ring.proj R A i r :=
+begin
+  conv_lhs { rw [←@graded_ring.right_inv R _ ι A _ _ _ r,
+    ←direct_sum.sum_support_of _ (@graded_ring.decompose R _ ι A _ _ _ r),
+    add_monoid_hom.map_sum], },
+  unfold graded_ring.support,
+  unfold graded_ring.proj,
+  apply finset.sum_congr, ext, simp only [dfinsupp.mem_support_to_fun],
+  intros i hi, simp only [ne.def, dfinsupp.mem_support_to_fun, subtype.val_eq_coe] at hi ⊢,
+  unfold direct_sum.add_subgroup_coe,
+  rw direct_sum.to_add_monoid_of, refl,
+end
+
+lemma graded_ring.mul_proj (r r' : R) (i : ι) :
+  graded_ring.proj R A i (r * r') =
+  ∑ ij in finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i)
+    ((graded_ring.support R A r).product (graded_ring.support R A r')),
+    (graded_ring.proj R A ij.1 r) * (graded_ring.proj R A ij.2 r') :=
+begin
+  have set_eq : (graded_ring.support R A r).product (graded_ring.support R A r') =
+  finset.filter (λ ij : ι × ι, ij.1 + ij.2 = i) _ ∪
+  finset.filter (λ ij : ι × ι, ij.1 + ij.2 ≠ i) _ := (finset.filter_union_filter_neg_eq _ _).symm,
+  conv_lhs { rw [graded_ring.as_sum R A r, graded_ring.as_sum R A r', finset.sum_mul_sum,
+    add_monoid_hom.map_sum, set_eq] },
+  rw finset.sum_union,
+  suffices : ∑ (x : ι × ι) in finset.filter (λ (ij : ι × ι), ij.fst + ij.snd ≠ i)
+    ((graded_ring.support R A r).product (graded_ring.support R A r')),
+  (graded_ring.proj R A i) ((graded_ring.proj R A x.fst) r * (graded_ring.proj R A x.snd) r') = 0,
+  rw [this, add_zero], apply finset.sum_congr rfl,
+  rintros ⟨j, k⟩ h, simp only [finset.mem_filter, finset.mem_product] at h ⊢,
+  obtain ⟨⟨h₁, h₂⟩, h₃⟩ := h,
+  rw ←h₃,
+  obtain ⟨a, rfl⟩ := (graded_ring.recompose R A).bijective.surjective r,
+  obtain ⟨b, rfl⟩ := (graded_ring.recompose R A).bijective.surjective r',
+  rw [graded_ring.proj_recompose, graded_ring.proj_recompose, ←ring_equiv.map_mul,
+    direct_sum.of_mul_of, graded_ring.proj_recompose],
+  congr, rw [direct_sum.of_eq_same],
+
+  apply finset.sum_eq_zero, rintros ⟨j, k⟩ h,
+  simp only [ne.def, finset.mem_filter, finset.mem_product] at h ⊢,
+  obtain ⟨⟨h₁, h₂⟩, h₃⟩ := h,
+  obtain ⟨a, rfl⟩ := (graded_ring.recompose R A).bijective.surjective r,
+  obtain ⟨b, rfl⟩ := (graded_ring.recompose R A).bijective.surjective r',
+  rw [graded_ring.proj_recompose, graded_ring.proj_recompose, ←ring_equiv.map_mul,
+    direct_sum.of_mul_of, graded_ring.proj_recompose, direct_sum.of_eq_of_ne],
+  simp only [ring_equiv.map_zero, add_monoid_hom.map_zero], intro rid, exact h₃ rid,
+
+  rw [finset.disjoint_iff_inter_eq_empty, finset.eq_empty_iff_forall_not_mem],
+  rintros ⟨j, k⟩ rid,
+  simp only [ne.def, finset.mem_filter, finset.mem_inter, finset.mem_product] at rid,
+  rcases rid with ⟨⟨_, h₁⟩, ⟨_, h₂⟩⟩, exact h₂ h₁,
+end
+
+lemma graded_ring.proj_homogeneous_element {x : R} {i : ι} (hx : x ∈ A i) :
+  graded_ring.proj R A i x = x :=
```

Comment:
I was able to prove this much more easily as something like:
```lean
lemma graded_algebra.proj_homogeneous_element {x : A} {i : ι} (hx : x ∈ 𝒜 i) :
  graded_algebra.proj 𝒜 i x = x :=
begin
  rw [graded_algebra.proj_apply, ←subtype.coe_mk x hx, subtype.coe_injective.eq_iff,
    ←graded_algebra.decompose_symm_of, alg_equiv.apply_symm_apply, direct_sum.of_eq_same],
end
```

Reply:
This is much nicer!